File: Demo/ZIKRouterDemoTests/ZIKViewRouterPerformTests.m
//
//  ZIKRouterTests.m
//  ZIKRouterTests
//
//  Created by zuik on 2018/4/11.
//  Copyright © 2018年 zuik. All rights reserved.
//

#import <XCTest/XCTest.h>
#import "AppRouteRegistry.h"
@import ZIKRouter;
@import ZIKRouter.Internal;
#import "SourceViewRouter.h"
#import "AViewInput.h"
#import "AViewRouter.h"

@interface ZIKRouterTests : XCTestCase
@property (nonatomic, strong) UIViewController *masterViewController;
@property (nonatomic, strong) SourceViewRouter *sourceRouter;
@property (nonatomic, strong) XCTestExpectation *leaveSourceViewExpectation;
@property (nonatomic, strong) XCTestExpectation *leaveTestViewExpectation;
@property (nonatomic, weak) ZIKDestinationViewRouter(id<AViewInput>) *router;
@property (nonatomic) ZIKViewRouteType routeType;
@end

@implementation ZIKRouterTests

#if !AUTO_REGISTER_ROUTERS

+ (void)load {
    [AViewRouter registerRoutableDestination];
    [SourceViewRouter registerRoutableDestination];
}

#endif

- (void)setUp {
    [super setUp];
    NSAssert(self.sourceRouter == nil, @"Last test didn't leave source view controler");
    NSAssert(self.router == nil, @"Last test didn't leave test view");
    if (self.masterViewController == nil) {
        UISplitViewController *root = (UISplitViewController *)[UIApplication sharedApplication].keyWindow.rootViewController;
        XCTAssertTrue([root isKindOfClass:[UISplitViewController class]]);
        UINavigationController *navigationController = [root.viewControllers firstObject];
        XCTAssertTrue([navigationController isKindOfClass:[UINavigationController class]]);
        self.masterViewController = [navigationController.viewControllers firstObject];
    }
    
    self.routeType = ZIKViewRouteTypePresentModally;
    self.leaveTestViewExpectation = [self expectationWithDescription:@"Remove test View Controller"];
}

- (void)tearDown {
    [super tearDown];
    self.leaveSourceViewExpectation = nil;
    self.leaveTestViewExpectation = nil;
}

- (void)enterTest:(void(^)(UIViewController *source))testBlock {
    [self enterSourceViewWithSuccess:testBlock];
}

- (void)enterSourceViewWithSuccess:(void(^)(UIViewController *source))successHandler {
    XCTestExpectation *expectation = [self expectationWithDescription:@"Show source View Controller"];
    self.leaveSourceViewExpectation = [self expectationWithDescription:@"Remove source View Controller"];
    self.sourceRouter = [SourceViewRouter performFromSource:self.masterViewController configuring:^(ZIKViewRouteConfig * _Nonnull config) {
        config.routeType = ZIKViewRouteTypePush;
        config.animated = NO;
        config.successHandler = ^(id  _Nonnull destination) {
            NSLog(@"%@: enterSourceView succeed", destination);
            [expectation fulfill];
            if (successHandler) {
                successHandler(destination);
            }
        };
    }];
}

- (void)leaveSourceView {
    [self.sourceRouter removeRouteWithConfiguring:^(ZIKViewRemoveConfiguration * _Nonnull config) {
        config.animated = NO;
        config.successHandler = ^{
            NSLog(@"LeaveSourceView succeed");
            [self.leaveSourceViewExpectation fulfill];
        };
    }];
    self.sourceRouter = nil;
}

- (void)leaveTestViewWithCompletion:(void(^)(BOOL success, ZIKRouteAction  _Nonnull routeAction, NSError * _Nullable error))completion {
    XCTAssertNotNil(self.router);
    [self.router removeRouteWithConfiguring:^(ZIKViewRemoveConfiguration * _Nonnull config) {
        config.successHandler = ^{
            NSLog(@"leaveTestView succeed");
            [self.leaveTestViewExpectation fulfill];
        };
        config.completionHandler = completion;
    }];
    self.router = nil;
}

- (void)leaveTest {
    if (self.router == nil) {
        [self leaveSourceView];
        return;
    }
    [self leaveTestViewWithCompletion:^(BOOL success, ZIKRouteAction  _Nonnull routeAction, NSError * _Nullable error) {
        [self leaveSourceView];
    }];
}

- (void)testPerformWithCompletionHandler {
    XCTestExpectation *expectation = [self expectationWithDescription:@"completionHandler"];
    {
        [self enterTest:^(UIViewController *source) {
            self.router = [ZIKRouterToView(AViewInput) performFromSource:source configuring:^(ZIKViewRouteConfiguration * _Nonnull config) {
                config.routeType = self.routeType;
                config.completionHandler = ^(BOOL success, id  _Nullable destination, ZIKRouteAction  _Nonnull routeAction, NSError * _Nullable error) {
                    XCTAssertTrue(success);
                    if (success) {
                        [expectation fulfill];
                    }
                    [self leaveTest];
                };
            }];
        }];
    }
    
    [self waitForExpectationsWithTimeout:5 handler:^(NSError * _Nullable error) {
        
    }];
}

- (void)testPerformWithSuccess {
    XCTestExpectation *expectation = [self expectationWithDescription:@"completionHandler"];
    {
        [self enterTest:^(UIViewController *source) {
            self.router = [ZIKRouterToView(AViewInput) performFromSource:source configuring:^(ZIKViewRouteConfiguration * _Nonnull config) {
                config.routeType = self.routeType;
                config.successHandler = ^(id  _Nonnull destination) {
                    XCTAssertNotNil(destination);
                    [expectation fulfill];
                    [self leaveTest];
                };
            }];
        }];
    }
    
    [self waitForExpectationsWithTimeout:5 handler:^(NSError * _Nullable error) {
        
    }];
}

- (void)testPerformWithPrepareDestination {
    XCTestExpectation *expectation = [self expectationWithDescription:@"completionHandler"];
    {
        [self enterTest:^(UIViewController *source) {
            self.router = [ZIKRouterToView(AViewInput) performFromSource:source configuring:^(ZIKViewRouteConfiguration * _Nonnull config) {
                config.routeType = self.routeType;
                config.prepareDestination = ^(id<AViewInput>  _Nonnull destination) {
                    destination.title = @"test title";
                };
                config.successHandler = ^(id<AViewInput>  _Nonnull destination) {
                    XCTAssertNotNil(destination);
                    XCTAssert([destination.title isEqualToString:@"test title"]);
                    [expectation fulfill];
                    [self leaveTest];
                };
            }];
        }];
    }
    
    [self waitForExpectationsWithTimeout:5 handler:^(NSError * _Nullable error) {
        
    }];
}

@end

File: Demo/ZIKViewRouterTests/ZIKViewRouterPerformTests.swift
//
//  ZIKViewRouterPerformTests.swift
//  ZIKViewRouterTests
//
//  Created by zuik on 2018/4/23.
//  Copyright © 2018 zuik. All rights reserved.
//

import XCTest
import ZRouter

extension ZIKViewRouterTestCase {
    func path(from source: UIViewController) ->ViewRoutePath {
        switch self.routeType {
        case .push:
            return .push(from: source)
        case .presentModally:
            return .presentModally(from: source)
        case .presentAsPopover:
            return .presentAsPopover(from: source)
        case .performSegue:
            return .performSegue(from: source)
        case .show:
            return .show(from: source)
        case .showDetail:
            return .showDetail(from: source)
        case .addAsChildViewController:
            return .addAsChildViewController(from: source)
        case .addAsSubview:
            return .addAsSubview(from: source.view)
        case .custom:
            return .custom(from: source)
        case .makeDestination:
            return .makeDestination
        }
    }
}

class ZIKViewRouterPerformTests: ZIKViewRouterTestCase {
    
    override func setUp() {
        super.setUp()
        self.routeType = .presentModally
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func configure(routeConfiguration config: ViewRouteConfig, source: ZIKViewRouteSource?) {
        config.animated = true
        config.routeType = self.routeType
    }
    
    func testPerformWithPrepareDestination() {
        let expectation = self.expectation(description: "prepareDestination")
        enterTest { (source) in
            self.router = Router.perform(
                to: RoutableView<AViewInput>(),
                path: self.path(from: source),
                configuring: { (config, prepareDest, _) in
                    self.configure(routeConfiguration: config, source: source)
                    prepareDest({ destination in
                        destination.title = "test title"
                        expectation.fulfill()
                    })
                    config.successHandler = { destination in
                        XCTAssert(destination is AViewInput)
                        XCTAssert((destination as! AViewInput).title == "test title")
                        self.handle({
                            XCTAssert(self.router?.state == .routed)
                            self.leaveTest()
                        })
                    }
            })?.router
        }
        waitForExpectations(timeout: 5, handler: { if let error = $0 {print(error)}})
    }
    
    func testPerformWithSuccessCompletionHandler() {
        let expectation = self.expectation(description: "completionHandler")
        enterTest { (source) in
            self.router = Router.perform(
                to: RoutableView<AViewInput>(),
                path: self.path(from: source),
                configuring: { (config, prepareDest, _) in
                    self.configure(routeConfiguration: config, source: source)
                    config.completionHandler = { (success, destination, action, error) in
                        XCTAssertTrue(success)
                        XCTAssertNil(error)
                        expectation.fulfill()
                        self.handle({
                            XCTAssert(self.router?.state == .routed)
                            self.leaveTest()
                        })
                    }
            })?.router
        }
        waitForExpectations(timeout: 5, handler: { if let error = $0 {print(error)}})
    }
    
    func testPerformWithErrorCompletionHandler() {
        let expectation = self.expectation(description: "completionHandler")
        enterTest { (source) in
            self.router = Router.perform(
                to: RoutableView<AViewInput>(),
                from: nil,
                configuring: { (config, prepareDest, _) in
                    self.configure(routeConfiguration: config, source: source)
                    config.completionHandler = { (success, destination, action, error) in
                        XCTAssertFalse(success)
                        XCTAssertNotNil(error)
                        expectation.fulfill()
                        self.handle({
                            XCTAssert(self.router == nil || self.router?.state == .unrouted)
                            self.leaveTest()
                        })
                    }
            })?.router
        }
        waitForExpectations(timeout: 5, handler: { if let error = $0 {print(error)}})
    }
    
    func testPerformWithSuccessCompletion() {
        let expectation = self.expectation(description: "completionHandler")
        enterTest { (source) in
            self.router = Router.perform(
                to: RoutableView<AViewInput>(),
                path: ViewRoutePath.presentModally(from: source),
                completion: { (success, destination, action, error) in
                    XCTAssertTrue(success)
                    XCTAssertNil(error)
                    expectation.fulfill()
                    self.handle({
                        XCTAssert(self.router?.state == .routed)
                        self.leaveTest()
                    })
            })?.router
        }
        waitForExpectations(timeout: 5, handler: { if let error = $0 {print(error)}})
    }
}
